clash-lib/src/Clash/Backend/VHDL.hs

```diff
@@ -661,6 +687,26 @@ funDec _ t@(Product _ labels elTys) = Just
                        (\(s,e) -> "fromSLV" <>
                           parens ("islv" <> parens (int s <+> "to" <+> int e)))
 
+funDec _ t@(Sum _ _) = Just
+  ( "function" <+> "toSLV" <+> parens("value" <+> colon <+> "in" <+> qualTyName t) <+> "return std_logic_vector" <> semi <> line <>
+    "function" <+> "fromSLV" <+> parens ("slv" <+> colon <+> "in" <+> "std_logic_vector") <+> "return" <+> qualTyName t <> semi
+  , "function" <+> "toSLV" <+> parens ("value" <+> colon <+> "in" <+> qualTyName t) <+> "return std_logic_vector" <+> "is" <> line <>
+    "begin" <> line <>
+    indent 2
+      ( "return" <+> "std_logic_vector" <>
+        parens ("to_unsigned" <>
+          parens (qualTyName t <> "'pos(value)" <> comma <+> int (typeSize t))
+        )) <> semi <> line <>
+    "end" <> semi <> line <>
+    "function" <+> "fromSLV" <+> parens ("slv" <+> colon <+> "in" <+> "std_logic_vector") <+> "return" <+> qualTyName t <+> "is" <> line <>
+    "begin" <> line <>
+    indent 2
+      ( "return" <+> qualTyName t <> "'val" <>
+        parens ("to_integer" <>
+          parens ("unsigned" <> parens "slv"))) <> semi <> line <>
+    "end" <> semi
+  )
+
```

Comment:
The `’val` and `'pos` attributes are positional according to all “vhdl val attribute” and “vhdl pos attribute” google results.

The encoding shouldn’t matter, when synthesis tools decide to use e.g. a one-hot encoding it’s up to those tools to pick to correct logic for those attributes.

Reply:
Picking the wrong representation doesn't matter as any ports with an enum type would have it converted to `std_logic_vector`, so this is safe provided `fromSLV` and `toSLV` are really the inverse of each other. The spec says `'val` and `'pos` should count sequentially, so this _should_ be true in practice.

FWIW if we did change things in the future to allow enum values in ports, the `enum_encoding` attribute is also in the VHDL spec, so you would hope tools would respect it and use the encoding you set there. But this is EDA, so all bets are off...

Reply:
Okay, thanks both. Would we need to worry about performance? I.e., non-corresponding encodings would introduce muxes/demuxes.

Reply:
Worry about performance for which thing specifically?

Reply:
Well, say the tooling would pick a one-hot encoding for the enum. The toslv/fromslv boundaries would then introduce muxes/demuxes while they wouldn't with identical encoding.

Reply:
Hmm, if that ends up being a problem for someone they can still explicitly set `enum_encoding` with synthesis attributes to control the internal representation, or use custom reprs to avoid generating enums at all

Reply:
When that happens, people should use custom bit-encodings.

Reply:
That seems.. user unfriendly. We're potentially introducing a regression here (i.e. _Clash_ is suddendly generating code that introduces muxes and demuxes) and as a solution we instruct our users to use custom bit encodings.

Reply:
Why wouldn't we generate `enum_encoding` right now to prevent the problem all-together?

Reply:
What's the other solution? Force binary encodings for enums using the `enum_encoding` attribute?

Reply:
Presumably. So from:

https://www.intel.com/content/www/us/en/programmable/quartushelp/18.0/index.htm#hdl/vhdl/vhdl_file_dir_enum_encoding.htm

it looks like 

> By default, the Intel® Quartus® Prime software "one-hot" encodes all user-defined Enumeration Types.

so we'd always introduce muxes/demuxes for enums. However, setting it is also not ideal:

> Note: If an Enumeration Type represents the states of a state machine, using the enum_encoding attribute to specify a manual state encoding prevents the Compiler from recognizing state machines based on the Enumeration Type. Instead, the Compiler processes these state machines as "regular" logic using the encoding specified by the attribute, and they are not listed as state machines in the Report window for the project.  If you wish to control the encoding for a recognized state machine, use the State Machine Processing logic option or the syn_encoding attribute. For more information on properly implementing state machines in VHDL, refer to Implementing State Machines.

I'm not sure how serious that problem is.

Reply:
Looking at the Quartus _Recommended HDL Coding Styles_ guide, it seems to suggest that it would only infer state machines when you write them as processes anyway. So I don't think this is a concern on the basis I don't think we generate anything this would apply to anyway

Reply:
That's a good point, so that particular downside doesn't really apply then.

Perhaps we can put this behind a feature flag (with the default set to "enabled"). If users experience regressions they can use the flag to disable it. We could remove the flag again if it turns out no one noticed any regressions (although I guess that's hard to measure).